src/db_ledger.rs

```diff
@@ -0,0 +1,622 @@
+            }
+        }
+
+        // Commit Step: Atomic write both the metadata and the data
+        let mut batch = WriteBatch::default();
+        if should_write_meta {
+            batch.put_cf(self.meta_cf.handle(&self.db), &meta_key, &serialize(&meta)?)?;
+        }
+
+        let serialized_blob_data = &new_blob.data[..BLOB_HEADER_SIZE + new_blob.size()?];
+        batch.put_cf(self.data_cf.handle(&self.db), key, serialized_blob_data)?;
+        self.db.write(batch)?;
+        Ok(consumed_queue)
+    }
+
+    // Fill 'buf' with num_blobs or most number of consecutive
+    // whole blobs that fit into buf.len()
+    //
+    // Return tuple of (number of blob read, total size of blobs read)
+    pub fn get_blob_bytes(
+        &mut self,
+        start_index: u64,
+        num_blobs: u64,
+        buf: &mut [u8],
+    ) -> Result<(u64, u64)> {
+        let start_key = DataCf::key(DEFAULT_SLOT_HEIGHT, start_index);
+        let mut db_iterator = self.db.raw_iterator_cf(self.data_cf.handle(&self.db))?;
+        db_iterator.seek(&start_key);
+        let mut total_blobs = 0;
+        let mut total_current_size = 0;
+        for expected_index in start_index..start_index + num_blobs {
+            if !db_iterator.valid() {
+                if expected_index == start_index {
+                    return Err(Error::IO(io::Error::new(
+                        io::ErrorKind::NotFound,
+                        "Blob at start_index not found",
+                    )));
+                } else {
+                    break;
+                }
+            }
+
+            // Check key is the next sequential key based on
+            // blob index
+            let key = &db_iterator.key().expect("Expected valid key");
+            let index = DataCf::index_from_key(key)?;
+            if index != expected_index {
+                break;
+            }
+
+            // Get the blob data
+            let value = &db_iterator.value();
+
+            if value.is_none() {
+                break;
+            }
+
+            let value = value.as_ref().unwrap();
+            let blob_data_len = value.len();
+
+            if total_current_size + blob_data_len > buf.len() {
+                break;
+            }
+
+            buf[total_current_size..total_current_size + value.len()].copy_from_slice(value);
+            total_current_size += blob_data_len;
+            total_blobs += 1;
+
+            // TODO: Change this logic to support looking for data
+            // that spans multiple leader slots, once we support
+            // a window that knows about different leader slots
+            db_iterator.next();
+        }
+
+        Ok((total_blobs, total_current_size as u64))
+    }
+
+    fn is_blob_data_valid(blob: &Blob) -> bool {
+        if let Ok(blob_size) = blob.size() {
+            let blob_data = &blob.data[BLOB_HEADER_SIZE..BLOB_HEADER_SIZE + blob_size];
+            let entry: BincodeResult<Entry> = deserialize(blob_data);
```

Comment:
misnamed, I think.  you're asking whether the blob can be an entry, which is possibly beyond the purview of this module...

you might add a trait to what gets stored or just bail on this?  what are you defending against with this check?